core/src/main/java/hudson/util/RingBufferLogHandler.java

```diff
@@ -24,42 +24,31 @@
 
 package hudson.util;
 
-import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
-import java.lang.ref.SoftReference;
+import hudson.remoting.ProxyException;
 import java.util.AbstractList;
 import java.util.List;
-import java.util.Objects;
+import java.util.logging.Formatter;
 import java.util.logging.Handler;
-import java.util.logging.Level;
 import java.util.logging.LogRecord;
+import java.util.logging.SimpleFormatter;
+import jenkins.util.JenkinsJVM;
 
 /**
  * Log {@link Handler} that stores the log records into a ring buffer.
  *
  * @author Kohsuke Kawaguchi
  */
-@SuppressFBWarnings(
-        value = "RV_RETURN_VALUE_IGNORED_NO_SIDE_EFFECT",
-        justification = "to guard against potential future compiler optimizations")
 public class RingBufferLogHandler extends Handler {
 
     private static final int DEFAULT_RING_BUFFER_SIZE = Integer.getInteger(RingBufferLogHandler.class.getName() + ".defaultSize", 256);
 
-    private static final class LogRecordRef extends SoftReference<LogRecord> {
-        LogRecordRef(LogRecord referent) {
-            super(referent);
-        }
-    }
-
-    static {
-        // Preload the LogRecordRef class to avoid an ABBA deadlock between agent class loading and logging.
-        LogRecord r = new LogRecord(Level.INFO, "<preloading>");
-        // We call Objects.hash() to guard against potential future compiler optimizations.
-        Objects.hash(new LogRecordRef(r).get());
-    }
+    /**
+     * Just to access {@link Formatter#formatMessage} which is not {@code static} though it could have been.
+     */
+    private static final Formatter dummyFormatter = new SimpleFormatter();
```

Comment:
There is already a field for a `Formatter` in the `Handler` class, which we could populate with an instance of `SimpleFormatter` in the constructor. Unlike the solution presented here, where a single `SimpleFormatter` is used for all `RingBufferLogHandler`s, using the field in the parent class would result in a separate `SimpleFormatter` for each `RingBufferLogHandler`. I think the solution offered here is more space-efficient at the cost of an additional field and an assumption that `Formatter#formatMessage` will always remain effectively static. I think it would be safer to avoid the latter assumption, even though the alternative is slightly less space-efficient, just in case `Formatter#formatMessage` is no longer effectively static in a future version of Java.

Reply:
The small number of handlers makes the space efficiency unimportant. I chose to use a separate field more for clarity: we are only calling `formatMessage` and _not_ `format` like the super field would normally be used for.

Reply:
Your logic for choosing to use a separate field makes sense to me, but I still think it is risky to assume that `formatMessage` is effectively static and thus that a single instance can be shared. That assumption may have been safe in the days of Java 8, but I feel that in recent years the Java standard libraries are not set in stone as much as they used to be.